tools/deploy/helper.go

```diff
@@ -0,0 +1,442 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to compute instance metadatas: %v", err)
+	}
+	return metadatas, nil
+}
+
+// validateInstances validates instances derived from placement against
+// instances derived from deployment, ensuring there are no duplicate instances
+// and the instances derived from two sources match against each other.
+func (h helper) computeInstanceMetadatas(
+	placementInstances []services.PlacementInstance,
+	deploymentInstances []Instance,
+) (instanceMetadatas, error) {
+	if len(placementInstances) != len(deploymentInstances) {
+		errMsg := "number of instances is %d in the placement and %d in the deployment"
+		return nil, fmt.Errorf(errMsg, len(placementInstances), len(deploymentInstances))
+	}
+
+	// Populate instance metadata from placement information.
+	metadatas := make(instanceMetadatas, len(placementInstances))
+	unique := make(map[string]int)
+	for i, pi := range placementInstances {
+		id := pi.ID()
+		_, exists := unique[id]
+		if exists {
+			return nil, fmt.Errorf("instance %s not unique in the placement", id)
+		}
+		endpoint := pi.Endpoint()
+		apiEndpoint, err := h.toAPIEndpointFn(endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("unable to convert placement endpoint %s to api endpoint: %v", endpoint, err)
+		}
+		unique[id] = i
+		metadatas[i].PlacementID = id
+		metadatas[i].ShardSetID = pi.ShardSetID()
+		metadatas[i].APIEndpoint = apiEndpoint
+	}
+
+	// Populate instance metadata from deployment information.
+	for _, di := range deploymentInstances {
+		id := di.ID()
+		placementID, err := h.toPlacementIDFn(id)
+		if err != nil {
+			return nil, fmt.Errorf("unable to convert deployment instance id %s to placement instance id", id)
+		}
+		idx, exists := unique[placementID]
+		if !exists {
+			return nil, fmt.Errorf("instance %s is in deployment but not in placement", id)
+		}
+		if metadatas[idx].DeploymentID != "" {
+			return nil, fmt.Errorf("instance %s not unique in the deployment", id)
+		}
+		metadatas[idx].DeploymentID = id
+		metadatas[idx].Revision = di.Revision()
+	}
+
+	return metadatas, nil
+}
+
+func (h helper) placementInstances() ([]services.PlacementInstance, error) {
+	stagedPlacement, onStagedPlacementDoneFn, err := h.placementWatcher.ActiveStagedPlacement()
+	if err != nil {
+		return nil, err
+	}
+	defer onStagedPlacementDoneFn()
+
+	placement, onPlacementDoneFn, err := stagedPlacement.ActivePlacement()
+	if err != nil {
+		return nil, err
+	}
+	defer onPlacementDoneFn()
+
+	return placement.Instances(), nil
+}
+
+func filterByRevision(metadatas instanceMetadatas, revision string) instanceMetadatas {
+	filtered := make(instanceMetadatas, 0, len(metadatas))
+	for _, metadata := range metadatas {
+		if metadata.Revision == revision {
+			continue
+		}
+		filtered = append(filtered, metadata)
+	}
+	return filtered
+}
+
+// instanceMetadata contains instance metadata.
+type instanceMetadata struct {
+	PlacementID  string
```

Comment:
I wasn't sure what placementID meant until I looked at how the struct was being created/used, especially cause the staged placement handler made me think there could be multiple placements per service instance. 

How about renaming to `id`, or adding a comment to indicate it's the placement instance's ID.